core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1164,19 +1167,52 @@ private void deleteInternal(LockedInodePath inodePath, boolean replayed, long op
       throw new InvalidPathException(ExceptionMessage.DELETE_ROOT_DIRECTORY.getMessage());
     }
 
-    List<Inode<?>> delInodes = new ArrayList<>();
-    delInodes.add(inode);
+    List<Pair<AlluxioURI, Inode<?>>> delInodes = new LinkedList<>();
+    // Persisted directories which can safely be deleted recursively from the UFS
+    // Mount points are not deleted recursively as we need to preserve the directory itself
+    Map<AlluxioURI, Inode<?>> safeRecursiveUFSDeletes = new HashMap<>();
+
+    Pair<AlluxioURI, Inode<?>> inodePair =
+        new Pair<AlluxioURI, Inode<?>>(inodePath.getUri(), inode);
+    delInodes.add(inodePair);
+    if (inode.isPersisted() && !replayed && inode.isDirectory()
+        && !mMountTable.isMountPoint(inodePath.getUri()) && (deleteOptions.isUnchecked()
+            || isUFSDeleteSafe(inodePath.getInode(), inodePath.getUri()))) {
+      safeRecursiveUFSDeletes.put(inodePath.getUri(), inode);
+    }
 
     try (InodeLockList lockList = mInodeTree.lockDescendants(inodePath, InodeTree.LockMode.WRITE)) {
-      delInodes.addAll(lockList.getInodes());
+      // Traverse inodes top-down
+      for (Inode descendant : lockList.getInodes()) {
+        AlluxioURI descendantPath = mInodeTree.getPath(descendant);
+        Pair<AlluxioURI, Inode<?>> descendantPair =
+            new Pair<AlluxioURI, Inode<?>>(descendantPath, descendant);
+        delInodes.add(descendantPair);
+        if (descendant.isPersisted() && !replayed && descendant.isDirectory()
+            && !mMountTable.isMountPoint(descendantPath)) {
+          if (deleteOptions.isUnchecked()
```

Comment:
what should be the change here?. i populate `safeRecursiveDeletes` even with `unchecked` so that during inode deletion individual delete is not called if parent is in the list